File: src/option.rs
use std::string::String;

pub struct Options {
    pub unary_rpc_promise: bool,
    pub grpc_package: String,
}

impl Options {
    pub fn parse(raw: &str) -> Options {
        let mut grpc_package = "@grpc/grpc-js";
        let mut unary_rpc_promise = false;

        let parts = raw.split(",");

        for part in parts {
            let mut kv = part.trim().split("=");
            let key = kv.next();
            match key {
                Some("grpc_package") => {
                    grpc_package = kv.next().unwrap_or("Expected a value for grpc_package")
                },
                Some("unary_rpc_promise") => {
                    unary_rpc_promise = kv.next().unwrap_or("Expected a value for unary_rpc_promise") == "true"
                },  
                // just silently ignore
                Some(_) => {}
                None => {}
            };
        }

        Options {
            grpc_package: String::from(grpc_package),
            unary_rpc_promise
        }
    }
}

#[test]
fn should_parse_empty() {
    let opt = Options::parse("");
    assert_eq!(opt.grpc_package, "@grpc/grpc-js");
    assert_eq!(opt.unary_rpc_promise, false);
}

#[test]
fn should_parse_grpc_package() {
    let opt = Options::parse("grpc_package=mygrpcpackage");
    assert_eq!(opt.grpc_package, "mygrpcpackage");
}

#[test]
fn should_parse_unary_promise() {
    let opt = Options::parse("unary_rpc_promise=true");
    assert_eq!(opt.unary_rpc_promise, true);
}

#[test]
fn should_parse_nontruthy_unary_promise() {
    let opt = Options::parse("unary_rpc_promise=false");
    assert_eq!(opt.unary_rpc_promise, false);
}

#[test]
fn should_ignore_unk_options() {
    let opt = Options::parse("ukn=1,unary_rpc_promise=true");
    assert_eq!(opt.unary_rpc_promise, true);
}


#[test]
fn should_parse_and_override() {
    let opt = Options::parse("unary_rpc_promise=false , grpc_package=mygrpcpackage ,unary_rpc_promise=true");
    assert_eq!(opt.grpc_package, "mygrpcpackage");
    assert_eq!(opt.unary_rpc_promise, true);
}

#[test]
fn should_parse_an_evil_option() {
    let opt = Options::parse("= , grpc_package=mygrpcpackage ,unary_rpc_promise=true");
    assert_eq!(opt.grpc_package, "mygrpcpackage");
    assert_eq!(opt.unary_rpc_promise, true);
}

File: src/option.ts
export type Target = "node" | "web"


export interface Options {
  unary_rpc_promise: boolean;
  grpc_package: string;
  no_namespace: boolean;
  json_names: boolean;
  explicit_override: boolean;
  target: Target;
}

export function parse(raw?: string): Options {
  if (!raw) {
    raw = "";
  }
  const options: Options = {
    unary_rpc_promise: false,
    grpc_package: "@grpc/grpc-js",
    no_namespace: false,
    json_names: false,
    explicit_override: false,
    target: "node"
  };
  for (const raw_option of raw.split(",")) {
    let [k, v] = raw_option.split("=", 2);
    switch (k) {
      case 'unary_rpc_promise': options[k] = v != "false"; break;
      case 'grpc_package':      options[k] = v;            break;
      case 'no_namespace':      options[k] = v != "false"; break;
      case 'json_names':        options[k] = v != "false"; break;
      case 'explicit_override': options[k] = v != "false"; break;
      case 'target':            options[k] = v as Target;  break;
    }
  }
  return options;
}
